### ISSUE TITLE
[BUG][CML]System freeze after reload modules.

### ISSUE THREAD

**Issue by YvonneYang2** [thesofproject/linux#966]
**Describe the bug**
Modules Reloading failed when remove modules and reload again

**To Reproduce**
1. Scenario 1:  Unload module with script "sudo ./sof_remove.sh" , reload module with script "sudo ./sof_insert.sh"(frequency: 6/10)
2. Scenario 2: run script "sudo ./sof_bootone.sh" (frequency:10/10)

Scripts can be found in: 
[kmod_scripts.zip](https://github.com/thesofproject/linux/files/3210166/kmod_scripts.zip)

**Expected result**
Modules will unload and reload successfully

**Actual result**
Freeze after modules reload. SSH also not work.

**Environment**
Firmware: 5fa2426 https://github.com/thesofproject/sof/tree/master
Kernel: b58f448 https://github.com/thesofproject/linux/tree/release/sof-v5.0
Topology: file: tools/topology/sof-hda-generic.tplg, sof-hda-generic-2ch.tplg same as firmware
Kconfig: https://sof-ci-storage.sh.intel.com/daily/20190521/release/linux/topic/sof-dev/901cd2a5-2/config-daily-hda-20190521.txt
platform ：CML laptop 

**dmesg**
```
[  117.107749] sof-audio-pci 0000:00:1f.3: ipc tx: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.107937] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.107941] sof-audio-pci 0000:00:1f.3: ipc tx: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.107948] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  117.107962] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  117.108112] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.108116] sof-audio-pci 0000:00:1f.3: ipc tx: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.108292] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x50010000: GLB_COMP_MSG: SET_VALUE
[  117.108296] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[  117.108356] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[  117.108410] skl_hda_dsp_generic skl_hda_dsp_generic: Dropping the link to ehdaudio0D2
[  117.176502] ------------[ cut here ]------------
[  117.176504] kernfs: can not remove 'driver', no directory
[  117.176526] WARNING: CPU: 2 PID: 3572 at fs/kernfs/dir.c:1504 kernfs_remove_by_name_ns+0x7f/0x90
[  117.176527] Modules linked in: snd_soc_skl_hda_dsp snd_soc_hdac_hdmi snd_hda_codec_realtek snd_hda_codec_generic snd_soc_dmic sof_pci_dev(-) snd_sof_intel_hda_common snd_soc_hdac_hda snd_sof_intel_hda snd_hda_ext_core snd_hda_codec snd_hda_core sof_acpi_dev snd_sof_intel_byt snd_sof_intel_bdw snd_sof_xtensa_dsp snd_sof_intel_hsw snd_sof_intel_ipc snd_sof snd_soc_acpi_intel_match snd_soc_acpi snd_soc_da7213 snd_soc_rt5640 snd_soc_rt5651 snd_soc_rt5645 snd_soc_rt5670 snd_soc_rl6231 snd_soc_core snd_pcm ax88179_178a usbnet iwlmvm snd_hwdep x86_pkg_temp_thermal intel_powerclamp mei_me processor_thermal_device intel_lpss_pci iwlwifi intel_lpss mei intel_soc_dts_iosf mfd_core intel_pch_thermal int3403_thermal int340x_thermal_zone int3400_thermal acpi_thermal_rel efivarfs xhci_pci xhci_hcd [last unloaded: snd_pcm]
[  117.176556] CPU: 2 PID: 3572 Comm: rmmod Tainted: G        W         5.0.0-cml-validation-20190521-0-hda-sof-b58f448 #5.0.0
[  117.176557] Hardware name: Dell Inc. Inspiron 5490/, BIOS 86.1.4 04/08/2019
[  117.176559] RIP: 0010:kernfs_remove_by_name_ns+0x7f/0x90
[  117.176560] Code: 00 31 c0 5b 5d 41 5c c3 48 c7 c7 80 c9 86 93 e8 37 ca 7f 00 b8 fe ff ff ff 5b 5d 41 5c c3 48 c7 c7 50 23 62 93 e8 d1 ad e2 ff <0f> 0b b8 fe ff ff ff eb d0 0f 1f 84 00 00 00 00 00 41 57 41 56 41
[  117.176561] RSP: 0018:ffffa1690195fcf0 EFLAGS: 00010282
[  117.176561] RAX: 0000000000000000 RBX: ffff9321e5155828 RCX: 0000000000000006
[  117.176562] RDX: 0000000000000007 RSI: 0000000000000086 RDI: ffff9321e8295410
[  117.176562] RBP: ffffffff936adbd2 R08: 0000000000000040 R09: 0000000000000004
[  117.176563] R10: ffff9321bdb8c0b0 R11: 0000000000000001 R12: 0000000000000000
[  117.176563] R13: ffffffffc04ec780 R14: 0000000000000060 R15: 0000000000000000
[  117.176564] FS:  00007f147d903540(0000) GS:ffff9321e8280000(0000) knlGS:0000000000000000
[  117.176564] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  117.176565] CR2: 000055965ff92b60 CR3: 00000001310f2005 CR4: 00000000003606e0
[  117.176565] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  117.176565] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  117.176566] Call Trace:
[  117.176574]  device_release_driver_internal+0x127/0x240
[  117.176578]  bus_remove_device+0xe0/0x150
[  117.176580]  device_del+0x133/0x350
[  117.176587]  snd_hdac_device_unregister+0x19/0x30 [snd_hda_core]
[  117.176591]  snd_hdac_ext_bus_device_remove+0x2a/0x60 [snd_hda_ext_core]
[  117.176595]  hda_dsp_remove+0x26/0x110 [snd_sof_intel_hda_common]
[  117.176601]  snd_sof_device_remove+0x73/0x90 [snd_sof]
[  117.176603]  sof_pci_remove+0x10/0x20 [sof_pci_dev]
[  117.176608]  pci_device_remove+0x36/0xb0
[  117.176609]  device_release_driver_internal+0x180/0x240
[  117.176610]  driver_detach+0x48/0x80
[  117.176611]  bus_remove_driver+0x4c/0xc0
[  117.176614]  pci_unregister_driver+0x1d/0x90
[  117.176624]  __x64_sys_delete_module+0x155/0x240
[  117.176627]  do_syscall_64+0x43/0xf0
[  117.176635]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  117.176637] RIP: 0033:0x7f147d4231b7
[  117.176669] Code: 73 01 c3 48 8b 0d d1 8c 2c 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 b8 b0 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d a1 8c 2c 00 f7 d8 64 89 01 48
[  117.176670] RSP: 002b:00007ffe10a71d88 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
[  117.176671] RAX: ffffffffffffffda RBX: 00007ffe10a71de8 RCX: 00007f147d4231b7
[  117.176671] RDX: 000000000000000a RSI: 0000000000000800 RDI: 000056549784d7b8
[  117.176672] RBP: 000056549784d750 R08: 00007ffe10a70d01 R09: 0000000000000000
[  117.176672] R10: 00007f147d49fcc0 R11: 0000000000000206 R12: 00007ffe10a71fb0
[  117.176673] R13: 00007ffe10a7284b R14: 000056549784c260 R15: 000056549784d750
[  117.176674] ---[ end trace a2e733b4aaee3f77 ]---
[  117.176839] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1010f0f successful
[  117.176848] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1000f0f successful
[  117.176851] sof-audio-pci 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[  117.176977] sof-audio-pci 0000:00:1f.3: Turning i915 HDAC power off
[  117.176983] sof-audio-pci 0000:00:1f.3: Dropping the link to 0000:00:02.0
[  117.335900] ------------[ cut here ]------------
[  117.335905] kernel BUG at mm/slub.c:3952!
[  117.335919] invalid opcode: 0000 [#1] SMP NOPTI
[  117.335926] CPU: 5 PID: 3633 Comm: rmmod Tainted: G        W         5.0.0-cml-validation-20190521-0-hda-sof-b58f448 #5.0.0
[  117.335929] Hardware name: Dell Inc. Inspiron 5490/, BIOS 86.1.4 04/08/2019
[  117.335941] RIP: 0010:kfree+0x115/0x160
[  117.335945] Code: 89 da 4c 89 e6 e8 fb 0c c5 00 48 8b 45 00 48 85 c0 75 e4 e9 0e ff ff ff 49 8b 02 a9 00 00 01 00 75 0a 49 8b 42 08 a8 01 75 02 <0f> 0b 49 8b 02 31 f6 a9 00 00 01 00 74 05 41 0f b6 72 51 5b 5d 41
[  117.335949] RSP: 0018:ffffa16900a3be98 EFLAGS: 00010246
[  117.335954] RAX: ffffe827453d7a08 RBX: ffff9321e5150800 RCX: 00000000820000af
[  117.335957] RDX: 0000000000000000 RSI: 0000000000000001 RDI: 00006cdf00000000
[  117.335960] RBP: 0000000000000011 R08: 0000000000000000 R09: ffffffff924e8600
[  117.335963] R10: ffffe82745945400 R11: 0000000000000001 R12: ffffffff924e8649
[  117.335965] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[  117.335969] FS:  00007f589e3dd540(0000) GS:ffff9321e8340000(0000) knlGS:0000000000000000
[  117.335972] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  117.335975] CR2: 00007f589df79d21 CR3: 000000014f684006 CR4: 00000000003606e0
[  117.335978] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  117.335981] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  117.335983] Call Trace:
[  117.335995]  free_module+0x99/0x200
[  117.336003]  __x64_sys_delete_module+0x18d/0x240
[  117.336010]  do_syscall_64+0x43/0xf0
[  117.336018]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  117.336023] RIP: 0033:0x7f589defd1b7
[  117.336027] Code: 73 01 c3 48 8b 0d d1 8c 2c 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 b8 b0 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d a1 8c 2c 00 f7 d8 64 89 01 48
[  117.336030] RSP: 002b:00007ffe518a3138 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
[  117.336034] RAX: ffffffffffffffda RBX: 00007ffe518a3198 RCX: 00007f589defd1b7
[  117.336037] RDX: 000000000000000a RSI: 0000000000000800 RDI: 000055c88c2007b8
[  117.336040] RBP: 000055c88c200750 R08: 00007ffe518a20b1 R09: 0000000000000000
[  117.336042] R10: 00007f589df79cc0 R11: 0000000000000206 R12: 00007ffe518a3360
[  117.336045] R13: 00007ffe518a3846 R14: 000055c88c1ff260 R15: 000055c88c200750
[  117.336049] Modules linked in: snd_hda_ext_core(-) snd_hda_codec snd_hda_core sof_acpi_dev snd_sof_intel_byt snd_sof_intel_bdw snd_sof_xtensa_dsp snd_sof_intel_hsw snd_sof_intel_ipc snd_sof snd_soc_acpi_intel_match snd_soc_acpi snd_soc_da7213 snd_soc_rt5640 snd_soc_rt5651 snd_soc_rt5645 snd_soc_rt5670 snd_soc_rl6231 snd_soc_core snd_pcm ax88179_178a usbnet iwlmvm snd_hwdep x86_pkg_temp_thermal intel_powerclamp mei_me processor_thermal_device intel_lpss_pci iwlwifi intel_lpss mei intel_soc_dts_iosf mfd_core intel_pch_thermal int3403_thermal int340x_thermal_zone int3400_thermal acpi_thermal_rel efivarfs xhci_pci xhci_hcd [last unloaded: snd_hda_ext_core]
[  117.336082] ---[ end trace a2e733b4aaee3f78 ]---
[  117.336088] RIP: 0010:kfree+0x115/0x160
[  117.336092] Code: 89 da 4c 89 e6 e8 fb 0c c5 00 48 8b 45 00 48 85 c0 75 e4 e9 0e ff ff ff 49 8b 02 a9 00 00 01 00 75 0a 49 8b 42 08 a8 01 75 02 <0f> 0b 49 8b 02 31 f6 a9 00 00 01 00 74 05 41 0f b6 72 51 5b 5d 41
[  117.336095] RSP: 0018:ffffa16900a3be98 EFLAGS: 00010246
[  117.336099] RAX: ffffe827453d7a08 RBX: ffff9321e5150800 RCX: 00000000820000af
[  117.336102] RDX: 0000000000000000 RSI: 0000000000000001 RDI: 00006cdf00000000
[  117.336104] RBP: 0000000000000011 R08: 0000000000000000 R09: ffffffff924e8600
[  117.336107] R10: ffffe82745945400 R11: 0000000000000001 R12: ffffffff924e8649
[  117.336110] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[  117.336113] FS:  00007f589e3dd540(0000) GS:ffff9321e8340000(0000) knlGS:0000000000000000
[  117.336116] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  117.336119] CR2: 00007f589df79d21 CR3: 000000014f684006 CR4: 00000000003606e0
[  117.336121] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  117.336124] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  209.298233] snd_sof_intel_hda: Unknown symbol snd_hdac_ext_bus_device_init (err -2)
```

[reload.txt](https://github.com/thesofproject/linux/files/3202199/reload.txt)
[scenario2_dmesg.txt](https://github.com/thesofproject/linux/files/3202232/scenario2_dmesg.txt)

**Comment by bardliao** [comment]
This issue can also be seen on sof-dev https://github.com/thesofproject/linux/commit/901cd2a52b667b4915f6c50ef2eb7813058bd96c
Below is the dmesg log of ``sudo ./sof_bootone.sh``
[dmesg.txt](https://github.com/thesofproject/linux/files/3205076/dmesg.txt)

**Comment by wenqingfu** [comment]
@bardliao to bi-sect on sof-5.0

**Comment by stevyan** [comment]
This issue can also be seen on CFL RVP HDA.

**Test Recipe**
date: 2019-05-22
sof (master): 3e423f41
linux (topic/sof-dev): e3ec7cf1
topology: sof-hda-generic.tplg

**Comment by bardliao** [comment]
@YvonneYang2 @ClarexZhou I have to say that I spent a lot of time to reproduce this issue.
Firstly, I tested with https://github.com/thesofproject/sof/tree/master/tools/kmod_scripts and and I can see the same dmesg log as you reported here. So I think I can reproduce the issue at that moment. But I get failed result on all the commit id and the daily test binaries. However my system didn't freeze actually.  @ClarexZhou suggested me to use https://gitlab.devtools.intel.com/sof-ci/sofcitest/tree/master/sof-test-suites/stress-test/kmod_scripts for testing then. And I got success result on all commit id and the daily test binaries. I am very confused until I ran ./sof_remove.sh, ./sof_insert.sh twice in one boot. My system will only freeze after the second ./sof_remove.sh. Anyway, I found the issue is a regression by https://github.com/thesofproject/linux/commit/18e1f86730d35056c9a1ce92f0532a4106143d11

BTW, can we have the same kmod_scripts in both github and gitlab?

**Comment by plbossart** [comment]
> BTW, can we have the same kmod_scripts in both github and gitlab?

Why would they be different? WTH?

**Comment by ranj063** [comment]
@bardliao @plbossart on my CML (with HDA codec), I can see that "rmmod sof_pci_dev" results in a kernel panic. And this is what happens:

snd_sof_device_remove()
  -> unregister machine driver
      --> hda_dsp_remove
         --> snd_hdac_ext_bus_device_remove
             --> again try to unregister machine driver (I think)

So if I remove the machine driver unregister from snd_sof_device_remove(), rmmod succeeds. I'll run the kmod script to confirm my theory

**Comment by wenqingfu** [comment]
> BTW, can we have the same kmod_scripts in both github and gitlab?
> Why would they be different? WTH?

talked with @lgirdwood about this a few days ago, the test directory was incorporated into sof repo while we merge the sof-tools (IIRC) repo into sof.  There is nobody maintaining those files now. we will take the SOF test cases out as a separate project or move them to ALSA test suite. It's part of the CI development to-do.

**Comment by wenqingfu** [comment]
> I found the issue is a regression by [18e1f86](https://github.com/thesofproject/linux/commit/18e1f86730d35056c9a1ce92f0532a4106143d11)

this is #928

**Comment by ranj063** [comment]
> > I found the issue is a regression by [18e1f86](https://github.com/thesofproject/linux/commit/18e1f86730d35056c9a1ce92f0532a4106143d11)
> 
> this is #928

@bardliao are you sure this commit caused th regression? I reverted this and I still see the problem. The only thing that helped was not unregistering the machine driver. I will investigate more today.

**Comment by plbossart** [comment]
Yeah, I don't see how changing a kzalloc/kfree sequence to devm_kzalloc could cause the system to freeze. Either way, please run those module insert/remove scripts as part of your development process. They catch a lot of issues...

**Comment by bardliao** [comment]
> @bardliao are you sure this commit caused th regression? I reverted this and I still see the problem. The only thing that helped was not unregistering the machine driver. I will investigate more today.

@ranj063 Yes I double check it again. When I keep #928, I can see the "freeze" issue. And my system won't freeze after revert #928. Do you mean you can see the oops or your system is actually freezed  when you said you see the problem? I think the description of this issue is a little bit misleading. Actually, you can't see anything if you meet the freeze issue. i.e. Your system don't have any response at all. And you can always see below with or without #928 
```
[  796.994737] ------------[ cut here ]------------
[  796.994738] kernfs: can not remove 'driver', no directory
[  796.994744] WARNING: CPU: 1 PID: 4178 at fs/kernfs/dir.c:1504 kernfs_remove_by_name_ns+0x7f/0x90
[  796.994744] Modules linked in: snd_soc_skl_hda_dsp snd_soc_hdac_hdmi snd_hda_codec_realtek snd_hda_codec_generic snd_soc_dmic sof_pci_dev(-) snd_sof_intel_hda_common snd_soc_hdac_hda snd_sof_intel_hda snd_hda_ext_core snd_hda_codec snd_hda_core sof_acpi_dev snd_sof_intel_byt snd_soc_acpi_intel_match snd_sof_intel_bdw snd_sof_intel_ipc snd_sof snd_sof_xtensa_dsp snd_soc_acpi snd_soc_pcm512x_i2c snd_soc_pcm512x snd_soc_da7213 snd_soc_rt5640 snd_soc_rt5651 snd_soc_rt5645 snd_soc_rt5670 snd_soc_rl6231 snd_soc_core snd_pcm ax88179_178a usbnet snd_hwdep x86_pkg_temp_thermal intel_powerclamp intel_lpss_pci intel_lpss processor_thermal_device mei_me intel_soc_dts_iosf mfd_core mei intel_pch_thermal int3403_thermal int340x_thermal_zone int3400_thermal acpi_thermal_rel efivarfs xhci_pci xhci_hcd [last unloaded: snd_pcm]
[  796.994756] CPU: 1 PID: 4178 Comm: rmmod Tainted: G        W         5.1.0+ #1
[  796.994757] Hardware name: Dell Inc. Inspiron 7590/, BIOS 99.2.5[X02_T05] 04/03/2019
[  796.994758] RIP: 0010:kernfs_remove_by_name_ns+0x7f/0x90
[  796.994759] Code: 00 31 c0 5b 5d 41 5c c3 48 c7 c7 c0 cb 86 bd e8 a7 73 81 00 b8 fe ff ff ff 5b 5d 41 5c c3 48 c7 c7 08 63 60 bd e8 81 ec e1 ff <0f> 0b b8 fe ff ff ff eb d0 0f 1f 84 00 00 00 00 00 41 57 41 56 41
[  796.994759] RSP: 0018:ffff9f68447b7d10 EFLAGS: 00010286
[  796.994760] RAX: 0000000000000000 RBX: ffff97c0916c3800 RCX: 0000000000000006
[  796.994760] RDX: 0000000000000007 RSI: 0000000000000082 RDI: ffff97c09c055410
[  796.994761] RBP: ffffffffbd6925d6 R08: 00000000000017f6 R09: 0000000000000004
[  796.994761] R10: ffff97c06ab62980 R11: 0000000000000001 R12: ffffffffc02ae780
[  796.994762] R13: ffff97c0916c38a0 R14: 0000000000000000 R15: 0000000000000000
[  796.994762] FS:  00007ff8fff22540(0000) GS:ffff97c09c040000(0000) knlGS:0000000000000000
[  796.994763] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  796.994763] CR2: 0000562e199a34e0 CR3: 00000007f97c2002 CR4: 00000000003606e0
[  796.994764] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  796.994764] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  796.994764] Call Trace:
[  796.994768]  device_release_driver_internal+0x86/0x1d0
[  796.994769]  bus_remove_device+0xe0/0x150
[  796.994770]  device_del+0x156/0x370
[  796.994773]  snd_hdac_device_unregister+0x19/0x30 [snd_hda_core]
[  796.994774]  snd_hdac_ext_bus_device_remove+0x2a/0x60 [snd_hda_ext_core]
[  796.994776]  hda_dsp_remove+0x26/0x110 [snd_sof_intel_hda_common]
[  796.994778]  snd_sof_device_remove+0x73/0x90 [snd_sof]
[  796.994779]  sof_pci_remove+0x10/0x30 [sof_pci_dev]
[  796.994781]  pci_device_remove+0x36/0xb0
[  796.994782]  device_release_driver_internal+0xdf/0x1d0
[  796.994783]  driver_detach+0x36/0x70
[  796.994784]  bus_remove_driver+0x4c/0xc0
[  796.994785]  pci_unregister_driver+0x1d/0x90
[  796.994787]  __x64_sys_delete_module+0x155/0x240
[  796.994789]  do_syscall_64+0x43/0xf0
[  796.994791]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  796.994792] RIP: 0033:0x7ff8ffa451b7
[  796.994792] Code: 73 01 c3 48 8b 0d d1 8c 2c 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 b8 b0 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d a1 8c 2c 00 f7 d8 64 89 01 48
[  796.994793] RSP: 002b:00007fff951a4008 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
[  796.994794] RAX: ffffffffffffffda RBX: 00007fff951a4068 RCX: 00007ff8ffa451b7
[  796.994794] RDX: 000000000000000a RSI: 0000000000000800 RDI: 00005568c79027f8
[  796.994794] RBP: 00005568c7902790 R08: 00007fff951a2f81 R09: 0000000000000000
[  796.994795] R10: 00007ff8ffac1cc0 R11: 0000000000000206 R12: 00007fff951a4230
[  796.994795] R13: 00007fff951a5793 R14: 00005568c7901260 R15: 00005568c7902790
[  796.994796] ---[ end trace 8b3c7ebd9e281137 ]---
```

**Comment by bardliao** [comment]
> Yeah, I don't see how changing a kzalloc/kfree sequence to devm_kzalloc could cause the system to freeze. 

@plbossart I can't see it either so far. But I think we have to check when will the resource actually be free when we use devm_kzalloc. We know hdev will be free at snd_hdac_ext_bus_device_exit() if we use kzalloc. But I don't know when will it be free if we use devm_kzalloc so far.

**Comment by YvonneYang2** [comment]
Issue not reproduced after revert #928 with test recipe as below:

**Environment**
Rander UCM https://github.com/RanderWang/intel-UCM/commits/hdmi
Firmware: 7a78809 https://github.com/bardliao/sof/tree/dell-test     
Kernel: 1051f31 https://github.com/thesofproject/linux/tree/release/sof-v5.0  (**revert #928 for #966**)
Topology: file: tools/topology/sof-hda-generic.tplg, sof-hda-generic-2ch.tplg same as firmware
Platform ：CML laptop

**Comment by ranj063** [comment]
@bardliao @stevyan reverting #928 does not fix the problem.

Please try "rmmod sof_pci_dev" with and without the patch and you;ll see the exact same kernel oops. Running sof_bootone.sh seems to hide the problem without the patch

**Comment by libinyang** [comment]
If using devm mode, hdev will be freed before hdev->dev kobject is released. This is definitely wrong. I will continue debug why devm will cause the memory release sequence wrong.

**Comment by libinyang** [comment]
The bug is root caused. It's a complicated bug and it took me more than 1 week to root cause it.

Below is my finding.
1. When initialing the codecs, snd_hdac_ext_bus_device_init() will be called, and it will set hdev->dev.release = default_release.
However, for analog codec (not hdac_hdmi codec), hdac_hda_codec_probe() will be called later. And it will call snd_hda_codec_device_new(), which will reset codec->code.dev.release = snd_hda_codec_dev_release; This means hdac_hdmi's hdev dev release is default_release() defined in hdac_ext_bus.c, and other hda codec's hdev dev release is snd_hda_codec_dev_release().

Both default_release() and snd_hda_codec_dev_release() will call kfree() to free the hdac_device (or its container) in the current code.

2. When we run rmmod sof_pci_dev, it will free the sdev. If we use Struct hdac_device *hdev = devm_kzalloc(sdev->dev...). This means hdev will also be freed automatically.

In the removal, snd_hdac_ext_bus_device_remove() will be called to remove the hdev (in this function it is struct hdac_device *codec.
The name is not aligned in different places). 
However for hdac_hdmi, the hdev->dev is used by other code. 
So calling device.release() (the function default_release()) will be postponed. After after sdev is freed, the device.release() will be called. But for devm_xxx, hdev will also be freed when sdev is freed. This means hdev.dev after sdev is freed is invalid now as hdev has already freed. It will access invalid memory. This will cause the bug.

So I think we should not use devm_xxx, and let's free the hdev manually.

At the end of this topic, I still found 2 suspicious code in the current code.
1. in sound/soc/intel/skylate/skl.c
it calls hdev = devm_kazlloc() or hda_codec = devm_kzalloc().
As we will call kfree() in the current code, should we replace it with kzalloc()? Maybe we need cavs drivers owner's help on it.

2. in snd_hdac_ext_bus_device_remove()
It will call snd_hdac_device_unregister() to unregister the hdac_devices and put_device(&codec->dev) to release the dev.
For analog codec, snd_hdac_device_unregister()  will free the codec->dev's kobject. And snd_hda_codec_dev_release() will be called to free the hdac_device.
So it is invalid to call put_device(&codec->dev). If you print
refcound_read(&(codec->dev.kobj.kref.refcount)) for analog codec before put_device(), you will find the refcount has already been 0.

More details can be found:
https://mailman.alsa-project.org/pipermail/alsa-devel/2019-May/149640.html

**Comment by ranj063** [comment]
@libinyang i've already posted a solution for this bug in https://mailman.alsa-project.org/pipermail/alsa-devel/2019-May/150093.html and it has been accepted by Takashi.
Sorry didnt realize you were already working on this.

**Comment by libinyang** [comment]
@ranj063 I noticed your patch. And they are different bugs. But your patch did fix one of the issues I raised in the mail thread: https://mailman.alsa-project.org/pipermail/alsa-devel/2019-May/149640.html. Your patch can fix the unbalanced device refcount.

**Comment by ranj063** [comment]
@libinyang so what is this other bug you are referring to?

**Comment by libinyang** [comment]
> so what is this other bug you are referring to?

@ranj063 after applying the patch of: https://mailman.alsa-project.org/pipermail/alsa-devel/2019-May/149640.html.
reload modules will cause system hang.

**Comment by ranj063** [comment]
@libinyang I could not reproduce this problem on my CML. There are some harmless errors in the dmesg wrt volume get failing or no reply expected that cause the script to fail. 
But if I run sof_remove.sh and sof_insert.sh, it works just fine for me.

**Comment by plbossart** [comment]
On 6/6/19 6:08 PM, Ranjani Sridharan wrote:
> @libinyang <https://github.com/libinyang> I could not reproduce this 
> problem on my CML. There are some harmless errors in the dmesg wrt 
> volume get failing or no reply expected that cause the script to fail.
> But if I run sof_remove.sh and sof_insert.sh, it works just fine for me.

maybe a red-herring again, but which version of sof_remove/insert are 
being used? Last time I checked QA was using a different set of scripts 
and i didn't get confirmation that we converged on the same tools.

**Comment by ranj063** [comment]
@plbossart i'm using the upstream version from SOF master

**Comment by ranj063** [comment]
@plbossart apparently I wasnt using the latest upstream version. And I do hit the system freeze issue but I have a feeling this might be related to the ordering of the modules. I'll look into it

**Comment by ranj063** [comment]
@plbossart it looks like it is just a matter waiting long enough between removing and inserting modules.
With a 10s sleep between sof_remove.sh and sof_insert.sh and then again for the next sof_remove.sh, I have no problems at all. I have a feeling pulseaudio might have a role to play here. Could you please take a look at it as well?

**Comment by plbossart** [comment]
we already use a 6s delay to allow pulseaudio to go to sleep and the system to go to ideal but I don't mind raising this value to 10s. On my one RVP I also had to increase this to 10s.

**Comment by ranj063** [comment]
@plbossart OK let me investigate a bit more and get back with a recommendation

**Comment by libinyang** [comment]
patch is submitted to fix this issue.
That patch reverts commit 18e1f86.

**Comment by plbossart** [comment]
@libinyang This issue is going nowhere. I can't figure out what the problem since your diagnosis did not take into account Ranjani's patch upstream. The revert without explanation is not acceptable, sorry.

**Comment by ranj063** [comment]
@plbossart I think the right fix for this issue is below. I will submit a PR today after Libin tests it. I was able to pass 100 module unload/reload iterations with this. Libin was on the right track in terms of root causing but the revert is not the right solution.
```
diff --git a/sound/pci/hda/hda_codec.c b/sound/pci/hda/hda_codec.c
index ccb53f1c1e1a..f28ed294ee67 100644
--- a/sound/pci/hda/hda_codec.c
+++ b/sound/pci/hda/hda_codec.c
@@ -860,7 +860,12 @@ static void snd_hda_codec_dev_release(struct device *dev)
        snd_hda_sysfs_clear(codec);
        kfree(codec->modelname);
        kfree(codec->wcaps);
-       kfree(codec);
+       /*
+        * In the case of ASoC HD-audio bus, the hda_codec will be freed when the
+        * hdac_hda_priv pointer containing the hda_codec is freed.
+        */
+       if (codec->core.type == HDA_DEV_LEGACY)
+               kfree(codec);
 }
 
 #define DEV_NAME_LEN 31
```

**Comment by libinyang** [comment]
> @plbossart I think the right fix for this issue is below. I will submit a PR today after Libin tests it. I was able to pass 100 module unload/reload iterations with this. Libin was on the right track in terms of root causing but the revert is not the right solution.
> 
> ```
> diff --git a/sound/pci/hda/hda_codec.c b/sound/pci/hda/hda_codec.c
> index ccb53f1c1e1a..f28ed294ee67 100644
> --- a/sound/pci/hda/hda_codec.c
> +++ b/sound/pci/hda/hda_codec.c
> @@ -860,7 +860,12 @@ static void snd_hda_codec_dev_release(struct device *dev)
>         snd_hda_sysfs_clear(codec);
>         kfree(codec->modelname);
>         kfree(codec->wcaps);
> -       kfree(codec);
> +       /*
> +        * In the case of ASoC HD-audio bus, the hda_codec will be freed when the
> +        * hdac_hda_priv pointer containing the hda_codec is freed.
> +        */
> +       if (codec->core.type == HDA_DEV_LEGACY)
> +               kfree(codec);
>  }
>  
>  #define DEV_NAME_LEN 31
> ```

@ranj063 Your patch works. So I will close my PR.

**Comment by libinyang** [comment]
@YvonneYang2 As the bug is fixed by Ranjani, could we close it?

**Comment by YvonneYang2** [comment]
I think we should keep it open, reasons as below:

1. PR1026 is not been merged.
2. Issue still reproduced on ICL with latest topic/sof-dev branch.
3. Modules reload on CML is blocked by issue #1067. But we still tried to test it on CML with latest topic/sof-dev branch + PR1026, found that sof_pci_dev in lsmod will change to -1 during removing modules  after reload 2-3 times.

**Comment by YvonneYang2** [comment]
System froze during the second time modules removing. Issue reproduced on ICL and CML platforms.

Test recipe:
Firmware: 34d0db94 https://github.com/thesofproject/sof/commits/master
Kernel: 01373cf7 https://github.com/thesofproject/linux/commits/topic/sof-dev
Topology: file: tools/topology/sof-hda-generic-2ch.tplg same as firmware
platform： icl/CML

**Comment by keyonjie** [comment]
This issue happens at runtime suspended status, it has relationship with hda jack detection work. @RanderWang

**Comment by ranj063** [comment]
@keyonjie @libinyang @mengdonglin #1026 already fixes this issue. It has just been applied upstream today so should be merged with the next upstream merge

**Comment by keyonjie** [comment]
Thanks @ranj063 .
Hi @ClarexZhou @YvonneYang2 can you help check if #1026 can fix the issue you hit? if yes, let's close this issue, and report if any new issues happen.

**Comment by YvonneYang2** [comment]
As comment https://github.com/thesofproject/linux/issues/966#issuecomment-508653850 said

> 3. Modules reload on CML is blocked by issue #1067. 

It will always reproduce bug #1067 during check pr#1026 , so cannnot test it.

**Comment by ClarexZhou** [comment]
Error not occurred with both PR #1026 and #1097.
Kernel (): https://github.com/thesofproject/linux/tree/topic/module_reload. Commit  4be09a5
FW (Master): Commit 9d28da9c
Topology: sof-hda-generic.tplg. Same with FW.

**Comment by ClarexZhou** [comment]
Error not seen after both PR #1026 and #1097 merged.

linux (topic/sof-dev :8b8f4377
sof (master): f837e57b
Platform: ICL hda rvp, cml Hda notebook, CFL-S hda rvp
